src/editor/codemirror/dnd-decorations.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  Decoration,
+  DecorationSet,
+  EditorView,
+  ViewPlugin,
+  ViewUpdate,
+} from "@codemirror/view";
+import { RangeSetBuilder } from "@codemirror/rangeset";
+import { StateEffect } from "@codemirror/state";
+
+export const timeoutDropDecorationEffect = StateEffect.define<{}>({});
+
+const dndDecorationsViewPlugin = ViewPlugin.fromClass(
+  class {
+    decorations: DecorationSet;
+    previewLineFroms = new Set<number>();
+    droppedRecentLineFroms = new Set<number>();
+    droppedDoneLineFroms = new Set<number>();
+
+    constructor(view: EditorView) {
+      this.decorations = this.dndDecorationsForLines(view);
+    }
+
+    update(update: ViewUpdate) {
+      for (const transaction of update.transactions) {
+        if (
+          transaction.effects.find((e) => e.is(timeoutDropDecorationEffect))
+        ) {
+          this.droppedDoneLineFroms.clear();
+          this.droppedRecentLineFroms.forEach((v) =>
+            this.droppedDoneLineFroms.add(v)
+          );
+          this.previewLineFroms.clear();
+          this.droppedRecentLineFroms.clear();
+        }
+      }
+
```

Comment:
This is a bit messy and definitely needs testing.
Could we have an array sized to the visible lines or some other better structure?

Reply:
We can't reasonably do this but I've tidied up a bit while keeping the approach essentially the same.